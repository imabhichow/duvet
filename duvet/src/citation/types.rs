use super::idset::IdSet;

static_intern!(Types, Type, u8);

#[derive(Clone, Copy, Debug)]
pub struct TypeSet(IdSet);

impl TypeSet {
    pub fn enable(&mut self, ty: Type) {
        self.set(ty, true);
    }

    pub fn disable(&mut self, ty: Type) {
        self.set(ty, false);
    }

    pub fn set(&mut self, ty: Type, enabled: bool) {
        self.0.set(ty.0, enabled)
    }

    pub fn get(&self, ty: Type) -> bool {
        self.0.get(ty.0)
    }
}

impl core::iter::FromIterator<Type> for TypeSet {
    fn from_iter<T: IntoIterator<Item = Type>>(iter: T) -> Self {
        Self(iter.into_iter().map(|v| v.0).collect())
    }
}

static_intern!(Levels, Level, u8);
