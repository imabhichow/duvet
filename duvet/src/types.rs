attribute!(pub const CODE: ());
attribute!(pub const TEST: ());
attribute!(pub const FUNCTION: ());
attribute!(pub const EXECUTIONS: u64);

attribute!(pub const FATAL: ());
attribute!(pub const ERROR: ());
attribute!(pub const WARNING: ());
attribute!(pub const INFO: ());
attribute!(pub const DEBUG: ());
