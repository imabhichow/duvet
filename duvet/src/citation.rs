pub mod idset;
pub mod parse;
pub mod rangemap;
pub mod spec;
pub mod token;
pub mod tree;
pub mod types;
