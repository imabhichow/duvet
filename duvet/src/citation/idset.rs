use core::mem::size_of;

#[derive(Clone, Copy, Debug, Default)]
pub struct IdSet(u64);

impl IdSet {
    pub fn enable(&mut self, index: u8) {
        self.set(index, true);
    }

    pub fn disable(&mut self, index: u8) {
        self.set(index, false);
    }

    pub fn set(&mut self, index: u8, enabled: bool) {
        debug_assert!((index as usize) < size_of::<Self>() * 8);
        let flag = 1 << index;
        if enabled {
            self.0 |= flag;
        } else {
            self.0 &= !flag;
        }
    }

    pub fn get(&self, index: u8) -> bool {
        debug_assert!((index as usize) < size_of::<Self>() * 8);
        let flag = 1 << index;
        self.0 & flag != 0
    }
}

impl core::iter::FromIterator<u8> for IdSet {
    fn from_iter<T: IntoIterator<Item = u8>>(iter: T) -> Self {
        let mut set = IdSet::default();
        for index in iter {
            set.enable(index);
        }
        set
    }
}
