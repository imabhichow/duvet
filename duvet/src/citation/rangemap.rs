use std::{
    collections::{btree_map, BTreeMap},
    iter::{FromIterator, Peekable},
    marker::PhantomData,
    ops::Range,
};

#[derive(Clone, Debug)]
pub struct RangeMap<Offset: Copy + Ord, Id: Copy + Ord>(BTreeMap<(Offset, Id), bool>);

impl<Offset: Copy + Ord, Id: Copy + Ord> Default for RangeMap<Offset, Id> {
    fn default() -> Self {
        Self::new()
    }
}

impl<Offset: Copy + Ord, Id: Copy + Ord> RangeMap<Offset, Id> {
    pub fn new() -> Self {
        Self(BTreeMap::new())
    }

    pub fn insert(&mut self, range: Range<Offset>, id: Id) {
        // empty range don't need entries
        if range.start >= range.end {
            return;
        }

        self.insert_point(range.start, id, true);
        self.insert_point(range.end, id, false);
    }

    fn insert_point(&mut self, offset: Offset, id: Id, value: bool) {
        match self.0.entry((offset, id)) {
            btree_map::Entry::Occupied(entry) => {
                if *entry.get() != value {
                    entry.remove();
                }
            }
            btree_map::Entry::Vacant(entry) => {
                entry.insert(value);
            }
        }
    }

    pub fn iter<T: FromIterator<Id>>(&self) -> Iter<Offset, Id, T> {
        Iter {
            iter: self.0.iter().peekable(),
            stack: Stack::new(),
            t: PhantomData,
        }
    }
}

#[derive(Clone, Debug)]
pub struct Iter<'a, Offset: Copy + Ord, Id: Copy + Ord, T> {
    iter: Peekable<btree_map::Iter<'a, (Offset, Id), bool>>,
    stack: Stack<Id>,
    t: PhantomData<T>,
}

impl<'a, Offset, Id, T> Iterator for Iter<'a, Offset, Id, T>
where
    Offset: Copy + Ord,
    Id: Copy + Ord,
    T: FromIterator<Id>,
{
    type Item = (Range<Offset>, T);

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            if cfg!(debug_assertions) {
                // ensure we end up with an empty stack
                if self.iter.peek().is_none() {
                    assert!(self.stack.is_empty(), "invalid stack state");
                }
            }

            let stack = &mut self.stack;

            let (key, enabled) = self.iter.next()?;
            let (offset, id) = *key;
            stack.insert(id, *enabled);

            let mut end = offset;

            while self
                .iter
                .next_if(|(key, enabled)| {
                    let (next_offset, id) = **key;
                    let enabled = **enabled;

                    // find the exclusive end of the iteration
                    end = next_offset;

                    // collect all of the entries at the current offset
                    if next_offset == offset {
                        stack.insert(id, enabled);
                        return true;
                    }

                    // only keep going if the stack doesn't change sizes
                    stack.maybe_insert(id, enabled)
                })
                .is_some()
            {}

            // don't return empty lists
            if self.stack.is_empty() {
                continue;
            }

            let range = offset..end;
            let value = self.stack.ids().collect();

            return Some((range, value));
        }
    }
}

#[derive(Clone, Debug, Default)]
struct Stack<Id: Copy + Ord>(BTreeMap<Id, u32>);

impl<Id: Copy + Ord> Stack<Id> {
    fn new() -> Self {
        Stack(BTreeMap::new())
    }

    fn insert(&mut self, id: Id, enabled: bool) -> bool {
        if enabled {
            match self.0.entry(id) {
                btree_map::Entry::Occupied(mut entry) => {
                    let value = entry.get_mut();
                    *value += 1;
                    false
                }
                btree_map::Entry::Vacant(entry) => {
                    entry.insert(1);
                    true
                }
            }
        } else {
            match self.0.entry(id) {
                btree_map::Entry::Occupied(mut entry) => {
                    let value = entry.get_mut();
                    *value -= 1;

                    if *value == 0 {
                        entry.remove();
                        return true;
                    }

                    false
                }
                btree_map::Entry::Vacant(_) => {
                    debug_assert!(false, "invalid stack state");
                    false
                }
            }
        }
    }

    fn maybe_insert(&mut self, id: Id, enabled: bool) -> bool {
        if enabled {
            match self.0.entry(id) {
                btree_map::Entry::Occupied(mut entry) => {
                    let value = entry.get_mut();
                    *value += 1;
                    true
                }
                btree_map::Entry::Vacant(_) => {
                    // This would add a new entry
                    false
                }
            }
        } else {
            match self.0.entry(id) {
                btree_map::Entry::Occupied(mut entry) => {
                    let value = entry.get_mut();

                    if *value == 1 {
                        // this would remove an entry
                        false
                    } else {
                        *value -= 1;
                        true
                    }
                }
                btree_map::Entry::Vacant(_) => {
                    debug_assert!(false, "invalid stack state");
                    true
                }
            }
        }
    }

    fn ids(&self) -> impl Iterator<Item = Id> + '_ {
        self.0.keys().copied()
    }

    fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn map<const L: usize>(ranges: [(Range<u32>, u8); L]) -> Vec<(Range<u32>, Vec<u8>)> {
        let mut map = RangeMap::default();

        for (range, id) in ranges.iter() {
            map.insert(range.start..range.end, *id);
        }

        map.iter().collect()
    }

    macro_rules! snapshot_test {
        ($name:ident, [$(($range:expr, $id:expr)),* $(,)?]) => {
            #[test]
            fn $name() {
                insta::assert_debug_snapshot!(stringify!($name), map([$(($range, $id)),*]));
            }
        };
    }

    snapshot_test!(empty, []);
    snapshot_test!(single, [(0..4, 0)]);
    snapshot_test!(nested, [(0..10, 0), (4..5, 0)]);
    snapshot_test!(contiguous, [(0..4, 0), (4..10, 0)]);
    snapshot_test!(contiguous_reverse, [(4..10, 0), (0..4, 0)]);
    snapshot_test!(non_contiguous, [(0..4, 0), (6..10, 0)]);
    snapshot_test!(overlap, [(0..6, 0), (4..10, 0)]);
    snapshot_test!(multiple, [(0..10, 0), (4..5, 1)]);
    snapshot_test!(multiple_reverse, [(4..5, 1), (0..10, 0)]);
}
