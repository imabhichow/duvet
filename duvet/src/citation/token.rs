use duvet_core::fs::{ArcStr, Substr};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Token {
    Meta {
        location: Location,
        key: Substr,
        value: Substr,
    },
    UnnamedMeta {
        location: Location,
        value: Substr,
    },
    Content {
        location: Location,
        value: Substr,
    },
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Location {
    pub line: u32,
    pub indent: u32,
}

impl Location {
    pub fn new(line: usize, indent: usize) -> Self {
        debug_assert!(line < u32::MAX as usize);
        debug_assert!(indent < u32::MAX as usize);
        Self {
            line: line as _,
            indent: indent as _,
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Tokenizer<'a> {
    pub meta_prefix: &'a str,
    pub content_prefix: &'a str,
}

impl<'a> Tokenizer<'a> {
    pub fn tokenize(self, contents: &'a ArcStr) -> Iter {
        Iter::new(self, contents)
    }
}

pub struct Iter<'a> {
    tokenizer: Tokenizer<'a>,
    lines: core::str::Lines<'a>,
    contents: &'a ArcStr,
    lineno: usize,
}

impl<'a> Iter<'a> {
    fn new(tokenizer: Tokenizer<'a>, contents: &'a ArcStr) -> Self {
        Self {
            tokenizer,
            contents,
            lines: contents.lines(),
            lineno: 0,
        }
    }

    fn on_line(&mut self, line: &str) -> Option<Token> {
        self.lineno += 1;

        let total_len = line.len();
        let line = line.trim_start();
        if line.is_empty() {
            return None;
        }

        let indent = total_len - line.len();

        let location = Location::new(self.lineno, indent);

        if let Some(meta) = line.strip_prefix(&self.tokenizer.meta_prefix) {
            return Some(self.on_meta(meta, location));
        }

        if let Some(content) = line.strip_prefix(&self.tokenizer.content_prefix) {
            return Some(self.on_content(content, location));
        }

        None
    }

    fn on_content(&mut self, content: &str, location: Location) -> Token {
        let value = self.contents.substr_from(content);
        Token::Content { location, value }
    }

    fn on_meta(&mut self, meta: &str, location: Location) -> Token {
        let mut parts = meta.trim_start().splitn(2, '=');

        let key = parts.next().unwrap();
        let key = key.trim_end();
        let key = self.contents.substr_from(key);

        if let Some(value) = parts.next() {
            let value = value.trim_start();
            let value = self.contents.substr_from(value);
            Token::Meta {
                key,
                value,
                location,
            }
        } else {
            Token::UnnamedMeta {
                value: key,
                location,
            }
        }
    }
}

impl<'a> Iterator for Iter<'a> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            let line = self.lines.next()?;
            if let Some(token) = self.on_line(line) {
                return Some(token);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use duvet_core::fs::arcstr;

    macro_rules! snapshot_test {
        ($name:ident, $input:expr) => {
            snapshot_test!(
                $name,
                $input,
                Tokenizer {
                    meta_prefix: "//=",
                    content_prefix: "//#",
                }
            );
        };
        ($name:ident, $input:expr, $config:expr) => {
            #[test]
            fn $name() {
                let parser = $config;
                insta::assert_debug_snapshot!(
                    stringify!($name),
                    parser
                        .tokenize(&arcstr::literal!($input))
                        .collect::<Vec<_>>()
                );
            }
        };
    }

    snapshot_test!(empty, "");
    snapshot_test!(
        basic,
        r#"
        //= thing goes here
        //= meta=foo
        //= meta2 = bar
        //# content goes
        //# here
        "#
    );
    snapshot_test!(
        only_unnamed,
        r#"
        //= this is meta
        //= this is other meta
        "#
    );
    snapshot_test!(
        duplicate_meta,
        r#"
        //= meta=1
        //= meta=2
        "#
    );
    snapshot_test!(
        configured,
        r#"
        /*
         *= meta=goes here
         *# content goes here
         */
        "#,
        Tokenizer {
            meta_prefix: "*=",
            content_prefix: "*#",
        }
    );
}
