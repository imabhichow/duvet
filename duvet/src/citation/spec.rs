use crate::citation::{tree::Tree, types::Level};

#[derive(Clone, Debug)]
pub struct Requirement {
    level: Level,
    tree: Tree,
}
