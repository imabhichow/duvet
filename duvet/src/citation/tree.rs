use crate::citation::types::{Type, TypeSet, Types};
use arcstr::{ArcStr, Substr};
use std::{collections::VecDeque, sync::Arc};

#[derive(Clone, Debug)]
pub enum Tree {
    Type(Type),
    Any(Arc<[Tree]>),
    All(Arc<[Tree]>),
    Xor(Arc<[Tree]>),
    Not(Arc<Tree>),
}

impl Tree {
    pub fn query<Q: Query>(&self, mut query: Q) -> Q::Value {
        enum Event<'a> {
            Enter(&'a Tree),
            Exit(&'a Tree),
        }

        let mut stack = VecDeque::new();
        let mut values = vec![];
        stack.push_back(Event::Enter(self));

        while let Some(event) = stack.pop_front() {
            match event {
                Event::Enter(tree) => match tree {
                    Tree::Type(ty) => values.push(query.eval(*ty)),
                    Tree::Any(args) | Tree::All(args) | Tree::Xor(args) => {
                        stack.push_front(Event::Exit(tree));
                        for arg in args.iter().rev() {
                            stack.push_front(Event::Enter(arg));
                        }
                    }
                    Tree::Not(arg) => {
                        stack.push_front(Event::Exit(tree));
                        stack.push_front(Event::Enter(arg));
                    }
                },
                Event::Exit(tree) => {
                    macro_rules! call {
                        ($name:ident, $args:expr) => {{
                            let index = values.len() - $args.len();
                            let value = query.$name(&values[index..]);
                            let _ = values.drain(index..);
                            values.push(value);
                        }};
                    }
                    match tree {
                        Tree::Type(_) => unreachable!(),
                        Tree::Any(args) => call!(any, args),
                        Tree::All(args) => call!(all, args),
                        Tree::Xor(args) => call!(xor, args),
                        Tree::Not(_) => {
                            let arg = values.pop().expect("invalid stack state");
                            let value = query.not(arg);
                            values.push(value);
                        }
                    }
                }
            }
        }

        debug_assert_eq!(values.len(), 1);

        values.pop().expect("invalid stack state")
    }
}

pub trait Query {
    type Value;

    fn eval(&mut self, ty: Type) -> Self::Value;
    fn all(&mut self, args: &[Self::Value]) -> Self::Value;
    fn any(&mut self, args: &[Self::Value]) -> Self::Value;
    fn xor(&mut self, args: &[Self::Value]) -> Self::Value;
    fn not(&mut self, arg: Self::Value) -> Self::Value;
}

impl Query for &'_ TypeSet {
    type Value = bool;

    fn eval(&mut self, ty: Type) -> Self::Value {
        self.get(ty)
    }

    fn all(&mut self, args: &[Self::Value]) -> Self::Value {
        args.iter().all(|v| *v)
    }

    fn any(&mut self, args: &[Self::Value]) -> Self::Value {
        args.iter().any(|v| *v)
    }

    fn xor(&mut self, args: &[Self::Value]) -> Self::Value {
        let mut value = false;
        for arg in args.iter().copied() {
            if arg && value {
                return false;
            } else if arg {
                value = true;
            }
        }
        value
    }

    fn not(&mut self, arg: Self::Value) -> Self::Value {
        !arg
    }
}

#[derive(Clone, Copy, Debug)]
pub enum Op {
    Any,
    All,
    Xor,
    Not,
}

impl Op {
    fn close_parse(&self, mut args: Vec<Tree>) -> Tree {
        match self {
            Self::All => {
                assert!(!args.is_empty());
                if args.len() == 1 {
                    return args.pop().unwrap();
                }
                Tree::All(Arc::from(args))
            }
            Self::Any => {
                assert!(!args.is_empty());
                if args.len() == 1 {
                    return args.pop().unwrap();
                }
                Tree::Any(Arc::from(args))
            }
            Self::Xor => {
                assert!(!args.is_empty());
                if args.len() == 1 {
                    return args.pop().unwrap();
                }
                Tree::Xor(Arc::from(args))
            }
            Self::Not => {
                assert!(args.len() == 1);
                Tree::Not(Arc::new(args.pop().unwrap()))
            }
        }
    }
}

#[derive(Clone, Debug)]
pub enum Token {
    Any(Substr),
    All(Substr),
    Xor(Substr),
    Not(Substr),
    OpenParen(Substr),
    CloseParen(Substr),
    Type(Substr),
}

impl Token {
    pub fn iter(content: &ArcStr) -> TokenIter {
        TokenIter {
            content,
            cursor: &*content,
        }
    }
}

#[derive(Clone, Debug)]
pub struct TokenIter<'a> {
    content: &'a ArcStr,
    cursor: &'a str,
}

impl TokenIter<'_> {
    pub fn types(&self) -> Types {
        self.clone()
            .filter_map(|token| {
                if let Token::Type(ty) = token {
                    Some(ty)
                } else {
                    None
                }
            })
            .collect()
    }
}

impl<'a> Iterator for TokenIter<'a> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        let cursor = self.cursor.trim_start();

        if cursor.is_empty() {
            self.cursor = cursor;
            return None;
        }

        macro_rules! paren {
            ($pat:literal, $name:ident) => {
                if let Some(other) = cursor.strip_prefix($pat) {
                    self.cursor = other;
                    let token = &cursor[..1];
                    let token = self.content.substr_from(token);
                    let token = Token::$name(token);
                    return Some(token);
                }
            };
        }

        paren!("(", OpenParen);
        paren!(")", CloseParen);

        macro_rules! call {
            ($pat:ident, $name:ident) => {
                if let Some(other) = cursor.strip_prefix(stringify!($pat)) {
                    self.cursor = other;
                    let token = &cursor[..stringify!($pat).len()];
                    let token = self.content.substr_from(token);
                    let token = Token::$name(token);
                    return Some(token);
                }
            };
        }

        call!(ANY, Any);
        call!(ALL, All);
        call!(XOR, Xor);
        call!(NOT, Not);

        for (idx, ch) in cursor.char_indices() {
            if ch.is_alphanumeric() || ch == '_' || ch == '-' {
                continue;
            }

            if idx == 0 {
                todo!("invalid {:?}", ch);
            }

            let (token, cursor) = cursor.split_at(idx);
            self.cursor = cursor;
            let token = self.content.substr_from(token);
            let token = Token::Type(token);
            return Some(token);
        }

        let token = cursor;
        self.cursor = &cursor[..0];
        let token = self.content.substr_from(token);
        let token = Token::Type(token);
        Some(token)
    }
}

fn parse(mut tokens: TokenIter, types: &Types) -> Result<Tree, Token> {
    let mut state = State::new();

    #[derive(Debug)]
    struct State {
        ty: Op,
        args: Vec<Tree>,
        stack: Vec<(Op, Vec<Tree>)>,
    }

    impl State {
        fn new() -> Self {
            Self {
                ty: Op::All,
                args: vec![],
                stack: vec![],
            }
        }

        fn push(&mut self, arg: Tree) {
            self.args.push(arg);
        }

        fn call(&mut self, ty: Op) {
            let prev_type = core::mem::replace(&mut self.ty, ty);
            let prev_args = core::mem::take(&mut self.args);
            self.stack.push((prev_type, prev_args));
        }

        fn open(&mut self) {
            // TODO
        }

        fn close(&mut self) {
            let (prev_type, prev_args) = self.stack.pop().unwrap();
            let current_type = core::mem::replace(&mut self.ty, prev_type);
            let current_args = core::mem::replace(&mut self.args, prev_args);
            self.args.push(current_type.close_parse(current_args));
        }

        fn finish(self) -> Tree {
            self.ty.close_parse(self.args)
        }
    }

    for token in tokens {
        match token {
            Token::Any(_) => state.call(Op::Any),
            Token::All(_) => state.call(Op::All),
            Token::Xor(_) => state.call(Op::Xor),
            Token::Not(_) => state.call(Op::Not),
            Token::OpenParen(_) => {
                state.open();
            }
            Token::CloseParen(_) => {
                state.close();
            }
            Token::Type(ty) => {
                let ty = types.resolve(&ty).unwrap();
                state.push(Tree::Type(ty));
            }
        }
    }

    Ok(state.finish())
}

#[cfg(test)]
mod tests {
    use super::*;
    use arcstr::literal;

    fn tokenize(input: &'static str) -> Vec<Token> {
        Token::iter(&ArcStr::from(input)).collect()
    }

    fn parse(input: &'static str) -> Result<Tree, Token> {
        let input = ArcStr::from(input);
        let iter = Token::iter(&input);
        let types = iter.types();
        super::parse(iter, &types)
    }

    fn eval(expr: ArcStr, sets: &[(&[&'static str], bool)]) {
        let iter = Token::iter(&expr);
        let types = iter.types();
        let tree = super::parse(iter, &types).unwrap();

        for (set, result) in sets {
            let set: TypeSet = set
                .iter()
                .map(|item| types.resolve(item).expect("missing type"))
                .collect();

            assert_eq!(tree.query(&set), *result, "set = {:?}", set);
        }
    }

    macro_rules! test {
        ($name:ident, $input:expr, [$(($set:expr, $result:expr)),+ $(,)?]) => {
            #[test]
            fn $name() {
                insta::assert_debug_snapshot!(
                    concat!(stringify!($name), "__tokens"),
                    tokenize($input)
                );
                insta::assert_debug_snapshot!(
                    concat!(stringify!($name), "__tree"),
                    parse($input)
                );

                eval(literal!($input), &[
                    $(
                        (
                            &$set[..],
                            $result
                        ),
                    )+
                ][..]);
            }
        };
    }

    test!(simple, "citation", [([], false), (["citation"], true)]);
    test!(
        any,
        "ANY(citation test)",
        [
            ([], false),
            (["citation"], true),
            (["test"], true),
            (["citation", "test"], true),
        ]
    );
    test!(
        all,
        "ALL(citation test)",
        [
            ([], false),
            (["citation"], false),
            (["test"], false),
            (["citation", "test"], true),
        ]
    );
    test!(
        any_all,
        "ANY(ALL(citation test) exception)",
        [
            ([], false),
            (["citation"], false),
            (["test"], false),
            (["exception"], true),
            (["citation", "test"], true),
            (["citation", "test", "exception"], true),
        ]
    );
    test!(
        xor,
        "XOR(citation exception)",
        [
            ([], false),
            (["citation"], true),
            (["exception"], true),
            (["citation", "exception"], false),
        ]
    );
    test!(
        xor_any,
        "XOR(ANY(citation test) exception)",
        [
            ([], false),
            (["citation"], true),
            (["citation", "test"], true),
            (["exception"], true),
            (["citation", "exception"], false),
            (["test", "exception"], false),
        ]
    );
    test!(not, "NOT(citation)", [([], true), (["citation"], false)]);
}
