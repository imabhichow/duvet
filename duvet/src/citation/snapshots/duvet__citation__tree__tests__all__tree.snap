---
source: duvet/src/citation/tree.rs
assertion_line: 397
expression: "parse(\"ALL(citation test)\")"

---
Ok(
    All(
        [
            Type(
                Type(
                    0,
                ),
            ),
            Type(
                Type(
                    1,
                ),
            ),
        ],
    ),
)
